### PR TITLE
Adding support for @SchemaFieldDescription annotation that allows ann…

### PR DESCRIPTION
**Pull request by pabloem**
Adding support for @SchemaFieldDescription annotation that allows annotations on AutoSchema classes

The motivation here is to support descriptions for Typed Schema Transforms, but I can see this being valuable for anyone using autogenerated schemas, but wanting to get description data in them as well.

r: @reuvenlax 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by pabloem** [comment]
fyi @mahagen @manavgarg this will enable us to add descriptions to schematransform config fields

**Comment by pabloem** [comment]
Run Java PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java [hunk 1]
```diff
@@ -1121,7 +1121,8 @@ public boolean typesEqual(Field other) {
 
     private boolean equivalent(Field otherField, EquivalenceNullablePolicy nullablePolicy) {
       return getName().equals(otherField.getName())
-          && getType().equivalent(otherField.getType(), nullablePolicy);
+          && getType().equivalent(otherField.getType(), nullablePolicy)
+          && getDescription().equals(otherField.getDescription());
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:1125]: I'm not sure that description should be included here as it's not really part of the type.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueTypeInformation.java [hunk 2]
```diff
@@ -167,6 +174,15 @@ public static <T extends AnnotatedElement & Member> String getNameOverride(
     }
   }
 
+  public static <T extends AnnotatedElement & Member> String getFieldDescription(T member) {
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueTypeInformation.java:177]: return value should be @Nullable

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java (hunk 1 above)
Comment (pabloem) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:1125]: this is checking whole Field equivalence, not just type. For Type equivalence we have `FieldType.equivalent(..)`,  and `Field.typesEqual(...)` - thoughts?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueTypeInformation.java (hunk 2 above)
Comment (pabloem) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueTypeInformation.java]: done

**Review by mahagen — APPROVED** [review]
Thanks for taking this on. Are we also planning on supporting SchemaFieldTitles? For instance, providing the human-readable text "BigQuery source table" for "tableSpec"?

**Review by robertwb — COMMENTED** [review]
This looks very nice, and huge +1 to making it easier to provide this kind of documentation/metadata.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:1125]: Yes, it is meant to be used to check schema equivalence (where name + type is what matters). I don't think description should be semantic here.

**Review by manavgarg — APPROVED** [review]

**Comment by pabloem** [comment]
@reuvenlax I've changed it to check equality, not equivalence

**Comment by pabloem** [comment]
> Thanks for taking this on. Are we also planning on supporting SchemaFieldTitles? For instance, providing the human-readable text "BigQuery source table" for "tableSpec"?

We could support it as a schema option with another annotation (e.g. @SchemaOption<T>("optionName", OptionType<T>, optionValue<T>)  - thoughts from others? @reuvenlax

**Comment by pabloem** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by pabloem** [comment]
Run Java_Pulsar_IO_Direct PreCommit

**Comment by pabloem** [comment]
Run Java_Debezium_IO_Direct PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java (hunk 1 above)
Comment (pabloem) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java]: removed this and changed for an equality check.

**Comment by reuvenlax** [comment]
LGTM

**Comment by reuvenlax** [comment]
Run Java_HBase_IO_Direct PreCommit

**Pull request merged**